### PR TITLE
Implement as_strided using existing MPSGraph Shape APIs

### PR DESCRIPTION
**Pull request by chrisbbayley**
Replace gather side of as_strided mps implementation when possible with MPSGraph shape APIs gives notable perf gain over existing path, roughly equal to the perf gain from using a custom metal kernel. 

Correctness checked by running test_mps.py

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/View.mm [hunk 1]
```diff
@@ -19,10 +19,12 @@
...
 
 static std::string getStridedKey(const ScalarType& dtype, const IntArrayRef& base_shape,
-                          const IntArrayRef& new_shape, bool is_scatter)
+                                 const IntArrayRef& new_shape, const IntArrayRef& stride,
+                                 int64_t storage_offset, bool is_scatter)
 {
   return (is_scatter ? "scatter:" : "gather:") + getMPSTypeString(dtype) + "[" +
-         getArrayRefString(base_shape) + "]:[" + getArrayRefString(new_shape) + "]";
+         getArrayRefString(base_shape) + "]:[" + getArrayRefString(new_shape) + "]:[" +
+         getArrayRefString(stride) + "]:[" + to_string(storage_offset) + "]";
```
[the remark this answers is not in the document]
Comment (kulinseth) [reply aten/src/ATen/native/mps/operations/View.mm:27]: @chrisbbayley , do we need these keys to be unique.

[the remark this answers is not in the document]
Comment (chrisbbayley) [reply aten/src/ATen/native/mps/operations/View.mm:27]: Stride and storage offset change the graph generated, and cannot be tensor inputs. So these must be part of the key generation.

File: aten/src/ATen/native/mps/operations/View.mm [hunk 2]
```diff
@@ -78,6 +80,352 @@
...
+    // Skip for zero in dst shape
+    for (NSInteger dstDim = 0; dstDim < dstRank; dstDim++)
+      if (dstSizes[dstDim] == 0) { return nil; }
+  }
+
+  // 1. Flatten the inputTensor if neccessary
+  MPSGraphTensor *flatInputTensor = inputTensor;
+  {
+    // Flatten inputs to remove duplicate strides.
+    NSMutableArray *squeezeAxes = [[NSMutableArray alloc] init];
```
[the remark this answers is not in the document]
Comment (chrisbbayley) [reply aten/src/ATen/native/mps/operations/View.mm:243]: This is missing a release, I will add that.

File: aten/src/ATen/native/mps/operations/View.mm [hunk 3]
```diff
@@ -78,6 +80,352 @@
...
+  [missingSrcDims release];
+  free(dstDimToSliceLength);
+  free(dstDimToSliceOffset);
+
+  return broadcastTensor;
+}
+
+MPSGraphTensor* asStridedLayer_pattern(MPSGraph *graph, MPSGraphTensor *inputTensor, int dstRank, const IntArrayRef& dstSizes, const IntArrayRef& dstStrides, int offset) {
+#ifdef DEBUG_MPSGRAPH_SHAPE_API_AS_STRIDED
+  printf("input shape:");
```
[the remark this answers is not in the document]
Comment (chrisbbayley) [reply aten/src/ATen/native/mps/operations/View.mm:403]: Yes definitely, meant to remove these, will definitely do that.

File: aten/src/ATen/native/mps/operations/View.mm [hunk 4]
```diff
@@ -78,6 +80,352 @@
...
+                                          axes:squeezeAxes
+                                          name:nil];
+  }
+
+  int srcRank = (int)[[flatInputTensor shape] count];
+  NSDictionary *srcStrideToDimLengthOffset = getStrideToDimLengthOffsetDict(flatInputTensor, srcRank, offset);
+
+  // Populate the dimension order, slice info, and broadcast info
+  NSMutableArray *dstDimOrder = [[NSMutableArray alloc] init];
+  int *dstDimToSliceLength = (int *) malloc(sizeof(int) * dstRank);
```
[the remark this answers is not in the document]
Comment (chrisbbayley) [reply aten/src/ATen/native/mps/operations/View.mm]: Done

File: aten/src/ATen/native/mps/operations/View.mm (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisbbayley) [reply aten/src/ATen/native/mps/operations/View.mm]: Done

File: aten/src/ATen/native/mps/operations/View.mm (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisbbayley) [reply aten/src/ATen/native/mps/operations/View.mm:241]: Done

**Review by razarmehr — APPROVED** [review]
Looks good.

**Review by kulinseth — APPROVED** [review]
Looks good, thanks for the changes.

**Pull request merged**